crates/ide_db/src/symbol_index.rs

```diff
@@ -400,79 +433,352 @@ impl FileSymbolKind {
-    fn decl<N: HasName>(node: N) -> Option<(SmolStr, SyntaxNodePtr, TextRange)> {
-        let name = node.name()?;
-        let name_range = name.syntax().text_range();
-        let name = name.text().into();
-        let ptr = SyntaxNodePtr::new(node.syntax());
+    fn push_decl_macro(&mut self, macro_def: MacroDef) {
+        self.push_file_symbol(|s| {
+            let name = macro_def.name(s.db.upcast())?.as_text()?;
+            let source = macro_def.source(s.db.upcast())?;
 
-        Some((name, ptr, name_range))
+            let (ptr, name_ptr) = match source.value {
+                Either::Left(m) => {
+                    (SyntaxNodePtr::new(m.syntax()), SyntaxNodePtr::new(m.name()?.syntax()))
+                }
+                Either::Right(f) => {
+                    (SyntaxNodePtr::new(f.syntax()), SyntaxNodePtr::new(f.name()?.syntax()))
+                }
+            };
+
+            Some(FileSymbol {
+                name,
+                kind: FileSymbolKind::Macro,
+                container_name: s.current_container_name(),
+                loc: DeclarationLocation { hir_file_id: source.file_id, name_ptr, ptr },
+            })
+        })
     }
-    match_ast! {
-        match node {
-            ast::Fn(it) => decl(it),
-            ast::Struct(it) => decl(it),
-            ast::Enum(it) => decl(it),
-            ast::Trait(it) => decl(it),
-            ast::Module(it) => decl(it),
-            ast::TypeAlias(it) => decl(it),
-            ast::Const(it) => decl(it),
-            ast::Static(it) => decl(it),
-            ast::Macro(it) => decl(it),
-            ast::Union(it) => decl(it),
-            _ => None,
+
+    fn push_file_symbol(&mut self, f: impl FnOnce(&Self) -> Option<FileSymbol>) {
+        if let Some(file_symbol) = f(self) {
+            self.symbols.push(file_symbol);
         }
     }
 }
 
-fn to_file_symbol(node: &SyntaxNode, file_id: FileId) -> Option<FileSymbol> {
-    to_symbol(node).map(move |(name, ptr, name_range)| FileSymbol {
-        name,
-        kind: match node.kind() {
-            FN => FileSymbolKind::Function,
-            STRUCT => FileSymbolKind::Struct,
-            ENUM => FileSymbolKind::Enum,
-            TRAIT => FileSymbolKind::Trait,
-            MODULE => FileSymbolKind::Module,
-            TYPE_ALIAS => FileSymbolKind::TypeAlias,
-            CONST => FileSymbolKind::Const,
-            STATIC => FileSymbolKind::Static,
-            MACRO_RULES => FileSymbolKind::Macro,
-            MACRO_DEF => FileSymbolKind::Macro,
-            UNION => FileSymbolKind::Union,
-            kind => unreachable!("{:?}", kind),
-        },
-        range: node.text_range(),
-        ptr,
-        file_id,
-        name_range: Some(name_range),
-        container_name: None,
-    })
+#[cfg(test)]
+mod tests {
+
+    use base_db::fixture::WithFixture;
+    use expect_test::expect_file;
+
+    use super::*;
+
+    #[test]
+    fn test_symbol_index_collection() {
```

Comment:
Would be good to add something with inner items to the test fixture here

Reply:
What do you mean things with inner items?

Reply:
54fc98920b1f28b5a9ba42a67fcd2868cc0b614a maybe?

Reply:
Yep that's what I meant